server: ensure settings are up-to-date.

vrongmeal opened the pull request:
[WIP]

Context: #50271

Fixes #48005.

Release note: None

Signed-off-by: Vaibhav <vrongmeal@gmail.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55166)
<!-- Reviewable:end -->

knz commented:
@tbg this PR seems to "fill in the blank" of the PR you had started earlier #50271 - wanna have a look?

tbg commented:
Yep, I'll take a look! I had suggested this issue to @vrongmeal.

tbg reviewed (commented):
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @vrongmeal)

___
*[pkg/keys/doc.go, line 218 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MJ0zoab0-sJQsKZCmn5:-MJ0zoab0-sJQsKZCmn6:b6p5q5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/048c570ca1777e54608fc2bbf0054ed23953003d/pkg/keys/doc.go#L218)):*
> ```Go
> 	StoreIdentKey,               // "iden"
> 	StoreLastUpKey,              // "uptm"
> 	StoreSettingsKey,            // "stng"
> ```

Let's name this "cached settings" throughout, to make clear that it's about keeping a cached version of the settings (which in particular may not be up to date).

So: StoreCachedSettingsKey (stng is still fine), localStoreCachedSettings{Suffix,KeyMin,KeyMax}, {Decode,}StoreCachedSettingsKey(), TestDecodeStoreCachedSettingsKey(),  `{load,store}CachedSettingsKVs`, 
___
*[pkg/server/init.go, line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MJ10-Kb-eiM5J2iJR6I:-MJ10-Kb-eiM5J2iJR6J:b-anpugo) ([raw file](https://github.com/cockroachdb/cockroach/blob/048c570ca1777e54608fc2bbf0054ed23953003d/pkg/server/init.go#L162)):*
> ```Go
> }
> 
> // storeSettingsKVs stores settings locally. This helps in restoring the node
> ```

make a new file `settings_cache.go` and put it there.

Put a test into `settings_cache_test.go` that makes a small slice of `[]roachpb.KeyValue` (see `roachpb.MakeValueFromString()`) and store-loads-compares it.
___
*[pkg/server/init.go, line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MJ1-Ojs43sO87hyj3ES:-MJ1-Ojs43sO87hyj3ET:b2qgax3) ([raw file](https://github.com/cockroachdb/cockroach/blob/048c570ca1777e54608fc2bbf0054ed23953003d/pkg/server/init.go#L164)):*
> ```Go
> // storeSettingsKVs stores settings locally. This helps in restoring the node
> // restart with the at least the same settings with which it died.
> func storeSettingsKVs(ctx context.Context, eng storage.Engine, kvs []roachpb.KeyValue) error {
> ```

use a batch here for performance (writes to the engine sync, and you don't want to sync in a loop):

```go
b := eng.NewBatch()
for {} // ...
return b.Commit(false /* sync */)
```
___
*[pkg/server/init.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MJ1-fgQ5wZddfGzjhRW:-MJ1-fgQ5wZddfGzjhRX:bmho9sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/048c570ca1777e54608fc2bbf0054ed23953003d/pkg/server/init.go#L181)):*
> ```Go
> 		keys.LocalStoreSettingsKeyMin,
> 		keys.LocalStoreSettingsKeyMax,
> 		func(kv storage.MVCCKeyValue) (stop bool, err error) {
> ```

oh, thought you had already gotten all of these iterator type things in your previous work :-) I wouldn't be opposed to making this follow the same pattern in a separate PR, if you feel so inclined.


<!-- Sent from Reviewable.io -->

tbg commented:
> oh, thought you had already gotten all of these iterator type things in your previous work :-) I wouldn't be opposed to making this follow the same pattern in a separate PR, if you feel so inclined.

Heh, I guess I didn't merge that yet. Will do now.

Forgot to mention earlier: make sure to update this test as well:

https://github.com/cockroachdb/cockroach/blob/88db9f7cbd73b4c863bb698793b4d7b620d62f30/pkg/keys/printer_test.go#L35

tbg commented:
For updating the test, you're going to want to add something to https://github.com/cockroachdb/cockroach/blob/88db9f7cbd73b4c863bb698793b4d7b620d62f30/pkg/keys/printer.go#L185

to translate your "stng" (or whatever the abbreviation is) to `cachedSettings`.

vrongmeal commented:
> Heh, I guess I didn't merge that yet. Will do now.

Yeah I was going to say.

I'll make the rest of the changes.

tbg commented:
PS here are some more things you'll need to complete creation of that key: https://github.com/cockroachdb/cockroach/commit/36822e23a5d839e882850930ca6c77eca66fbce7

tbg commented:
(I just happen to also create a new key in my other work, so you can pretty much just follow along with that commit)

vrongmeal commented:
Great :)

vrongmeal committed:
keys: introduce CachedSettingsKey

These keys have a store-local prefix and contain settings for nodes
that are cached locally to restore when a node is restarted.

Release note: None

Signed-off-by: Vaibhav <vrongmeal@gmail.com>

tbg reviewed (commented):
> Reviewed 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @vrongmeal)



<!-- Sent from Reviewable.io -->

vrongmeal committed:
server: add methods to store and load cached settings

Methods {store,load}CachedSettingsKVs locally save and read
values from local store respectively. These can be used in
restoring the settings for the node which are at-least the
same settings with which it died.

Release note: None

Signed-off-by: Vaibhav <vrongmeal@gmail.com>

vrongmeal commented:
@tbg sorry for the delay. I rebased this on top of #55315 which also needs your review.

tbg reviewed (commented):
This looks good so far! I think the next step is to pull that refactor from https://github.com/cockroachdb/cockroach/pull/50271/files#diff-a7608f72d3f9b6840a0c1a4fc2d403baf141c0cf2cf7684561e6bd7940ac5c66R139 in that makes the settings updater stuff re-usable.

After that, we're ready to add a gossip callback that fires whenever the settings are updated (writing them to disk). This can be verified as working via an end-to-end test that sets up a test server, changes a cluster setting, and waits for the setting to show up on the disk.

Loading from disk will happen here: https://github.com/cockroachdb/cockroach/pull/50271/files#diff-06239ddccba92ce2c7f7913baaf56f2f3ea31e311df60b16dbc8332cce0fefb4R181 and we need to feed it to the settings updater as early as it can. I think we can do it right after `ServeAndWait` returns in `(*Server).PreStart()` (this is not what the prototype does btw).

> Reviewed 28 of 28 files at r3, 6 of 6 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-:-MJWo1XaDLfbafHnn5Uj:bhcjkys)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @vrongmeal)



<!-- Sent from Reviewable.io -->

vrongmeal committed:
server: ensure settings are up-to-date

In an ideal world a KV node would not declare itself as ready until
it has received the current cluster settings.

However, we cannot indiscriminately block on that because of the
chicken-and-egg problem that comes up when the node currently starting
is needed for quorum on the system config range. For example, if a three
node cluster is completely down, at least two of the nodes must be
online before current settings are received.

Instead do the following:

1. persist the settings locally on the first store whenever they are
   received, so that restarting nodes can come up with settings that are
   no staler than the ones they had when they went down.
2. if a new node joins the cluster, we can wait for the settings to show
   up (since the node is just joining, it is not required for any quorum).

Fixes #48005.

Release note: None

Signed-off-by: Vaibhav <vrongmeal@gmail.com>

vrongmeal commented:
Hey! I was out of town for a few days.
Anyways, I pulled the refactor from #50271. I also made other changes taking context from the same PR. A couple of tests fail as of now, still need to look into them. I think you can take a look at the code written so far.

tbg reviewed (commented):
I looked into the test failure using `TestClusterVersionUpgrade`. I noticed this error in the logs (run with `-show-logs`):

```
W201103 08:36:27.881845 341 server/settingsworker.go:139  [n1] error processing config KVs: while storing settings kvs: cannot have timestamp set in value on Put
```

due to this error, we never hit the `u.ResetRemaining()` just below the call to `storeCachedSettingsKVs`.

So two things to fix (and I pushed a fixup commit fixing them): failing to persist the settings shouldn't interfere with the rest, and we need to reset the timestamp before writing the cached settings. This is because we write the cached settings as non-MVCC values, so they need to have a zero timestamp.

> Reviewed 6 of 6 files at r6, 2 of 2 files at r7, 3 of 3 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-:-MLCFBlu-vyTsOg11nJm:b-9c30vw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg and @vrongmeal)

___
*[pkg/server/init.go, line 353 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MLC9oSV5KQ9PFgVVkcO:-MLC9oSW47u61-GmL5CT:bbyrszr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ea58ff7405e3b93f65d48d3ab7dbafec8dcf5ab/pkg/server/init.go#L353)):*
> ```Go
> 			s.mu.Unlock()
> 
> 			var initialSettingsKVs []roachpb.KeyValue
> ```

At this point, thankfully, joining the cluster via gossip is legacy behavior, so let's not put new code here. You can just revert this part of the diff.
We could consider returning the settings from the Join RPC (which replaces gossip-join), but let's hold off on that, at least in this PR, though it would be a good follow-up.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Looks good! Give this a rebase (you picked up a fairly flaky test on master) and then I'll give it one last look.

> Reviewed 11 of 11 files at r9, 2 of 2 files at r10, 4 of 4 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-:-MLIxCC2DtaKGDCx3Uu4:bnob6a7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @vrongmeal)

___
*[pkg/kv/kvserver/store.go, line 2174 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MLIvNuO9h54aMk4eo0o:-MLIvNuO9h54aMk4eo0p:b-lmvvn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/87becb02de8a284636eda963f33b2436e35b3483/pkg/kv/kvserver/store.go#L2174)):*
> // Cached cluster settings may be present on uninitialized engines.
___
*[pkg/server/settingsworker.go, line 133 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MLIwxYW4KUP1-6-9i17:-MLIwxYW4KUP1-6-9i18:btlazmz) ([raw file](https://github.com/cockroachdb/cockroach/blob/87becb02de8a284636eda963f33b2436e35b3483/pkg/server/settingsworker.go#L133)):*
> ```Go
> 				cfg := s.gossip.GetSystemConfig()
> 				u := s.st.MakeUpdater()
> 				// TODO(tbg): make sure s.engines[0] is initialized. Or rather, allow
> ```

Remove this TODO. The approach of writing to the engine no matter what and making sure that the cached settings get ignored for bootstrapping makes more sense than trying to ensure that an engine is initialized here (which I think isn't true, or you wouldn't have added that code to ignore cached settings during store init :-) )


<!-- Sent from Reviewable.io -->

vrongmeal commented:
Coincidentally, I just pushed the rebase. I'll make the changes you just suggested.

tbg commented:
I looked into whether applying the cached settings on server start could "clobber" a newer update received via gossip. The answer is no, but we should make this clearer/less prone to rot. I think the way to go is to pass the `initialSettingsKVs` to this method:

https://github.com/cockroachdb/cockroach/blob/9b8526454df4b9e43bb9d5e7005010f44cb4edac/pkg/server/server.go#L1574

That method sets up the gossip-based settings listener. By putting this method in charge of applying the initial cache to the settings before consulting gossip for updates, proper sequencing is "obvious".

Additionally, we should move that call above to the place where your PR currently populates the cache, https://github.com/cockroachdb/cockroach/blob/9b8526454df4b9e43bb9d5e7005010f44cb4edac/pkg/server/server.go#L1419

That way we continue to make sure that the settings are applied as early as possible in the start sequence. We will start the settings gossip event loop earlier, but that is not an issue.

Btw, you're catching some merge conflicts with some simplifications @irfansharif is landing these days, they shouldn't be substantial.

vrongmeal commented:
If I'm passing initial settings to
https://github.com/cockroachdb/cockroach/blob/9b8526454df4b9e43bb9d5e7005010f44cb4edac/pkg/server/server.go#L1574

I don't think we require 
https://github.com/cockroachdb/cockroach/blob/9b8526454df4b9e43bb9d5e7005010f44cb4edac/pkg/server/server.go#L1419

Moreover, I did not know if I should move `refreshSettings` above, and so I populated the initial settings before. Just confirming that we're on the same page.

tbg reviewed (commented):
Same page!

Almost there, looking good. Only have to move the loading of initial settings into inspectEngines.

> Reviewed 6 of 6 files at r12, 3 of 3 files at r13, 5 of 5 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-:-MLmaDA2DX-sTMpNiV63:b-8zz19j)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @vrongmeal)

___
*[pkg/server/init.go, line 199 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MLm_8_wDiF1jPBLEbB0:-MLm_8_wDiF1jPBLEbB1:bh3ykrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a744e1c18ac587e27add12ce94bec245a812f8c/pkg/server/init.go#L199)):*
> ```Go
> 	// If we're restarting an already bootstrapped node, return early.
> 	if s.inspectedDiskState.bootstrapped() {
> 		cachedSettings, err := loadCachedSettingsKVs(ctx, s.inspectedDiskState.initializedEngines[0])
> ```

Make this part of `inspectEngines`, so that we're never mutating an `initState` outside of that method.
___
*[pkg/server/server.go, line 1420 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MLmaS51CP4OyGVtifko:-MLmaS51CP4OyGVtifkp:bkg0btz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a744e1c18ac587e27add12ce94bec245a812f8c/pkg/server/server.go#L1420)):*
> ```Go
> 		return errors.Wrap(err, "during init")
> 	}
> ```

// Apply any cached initial settings (and start the gossip listener) as early as possible, to
// avoid spending time with stale settings.
___
*[pkg/server/settings_cache.go, line 37 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MLmakF25iYUdTW2OrS0:-MLmakF25iYUdTW2OrS1:b-h74w05) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a744e1c18ac587e27add12ce94bec245a812f8c/pkg/server/settings_cache.go#L37)):*
> ```Go
> 		}
> 	}
> 	return batch.Commit(false)
> ```

`batch.Commit(false /* sync */)`
___
*[pkg/server/settingsworker.go, line 125 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MLm_eB3AkIdF7PwroxY:-MLm_eB40V2CihKKtqti:bdf2954) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a744e1c18ac587e27add12ce94bec245a812f8c/pkg/server/settingsworker.go#L125)):*
> ```Go
> func (s *Server) refreshSettings(initialSettingsKVs []roachpb.KeyValue) error {
> 	ctx := s.AnnotateCtx(context.Background())
> 	// Since we have initial settings from ServeAndWait, we can process the
> ```

Just to avoid having to think about it, wrap this in an `if len(initialSettingsKVs) != 0 {}`. Comment could be something like

// If we have initial settings KV pairs loaded from the local engines, apply them before starting the gossip listener.


<!-- Sent from Reviewable.io -->

vrongmeal commented:
@tbg I had updated the PR yesterday with all the changes that you requested in your last review.

tbg reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> sorry about the long review cycles! You need to rebase, there's one last inconsequential comment, and then you can `bors r=tbg` the change.

bors delegate+

> Reviewed 7 of 7 files at r15, 3 of 3 files at r16, 5 of 5 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-:-MMGTIVT8cgysXHu4HXG:b-p2jy9l)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @vrongmeal)

___
*[pkg/server/server.go, line 787 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55166#-MMGSrVs7AH1ywjg0kC7:-MMGSrVs7AH1ywjg0kC8:b6u23ve) ([raw file](https://github.com/cockroachdb/cockroach/blob/d101162536534ca3764cfcab2a3c724a8d3c39ee/pkg/server/server.go#L787)):*
> ```Go
> 	}
> 	var initialSettingsKVs []roachpb.KeyValue
> 	if len(initializedEngines) != 0 {
> ```

There's a weird asymmetry here - we will persist them on uninitialized engines, but never load from them. Just use all of the engines here even though in practice it would not make a difference except in extraordinarily contrived cases (and even then the difference would not matter).


<!-- Sent from Reviewable.io -->

vrongmeal commented:
> Just use all of the engines here

Does this mean I should loop over all the engines (initialized and uninitialized) and append all the KVs to `initialSettingsKVs`?

tbg commented:
No, use the first one that has anything on it. I.e. loop and return when
something found.

On Tue, Nov 17, 2020, 01:41 Vaibhav <notifications@github.com> wrote:

> Just use all of the engines here
>
> Does this mean I should loop over all the engines (initialized and
> uninitialized) and append all the KVs to initialSettingsKVs?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/55166#issuecomment-728541740>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABGXPZHNAHA6KYRIJMELUYLSQHBDZANCNFSM4SBTEY3Q>
> .
>

vrongmeal commented:
bors r=tbg

vrongmeal commented:
> We could consider returning the settings from the Join RPC (which replaces gossip-join), but let's hold off on that, at least in this PR, though it would be a good follow-up.

@tbg should I start with this?

irfansharif commented on pkg/server/settings_cache.go:
```diff
@@ -0,0 +1,66 @@
+			return err
+		}
+	}
+	return batch.Commit(false /* sync */)
```
Drive by comment: I'm curious why we're choosing to not durably persist these writes here. I think we want to, no?

irfansharif commented on pkg/server/server.go:
```diff
@@ -1369,6 +1380,12 @@ func (s *Server) PreStart(ctx context.Context) error {
 		return errors.Wrap(err, "invalid init state")
 	}
 
+	// Apply any cached initial settings (and start the gossip listener) as early
+	// as possible, to avoid spending time with stale settings.
```
> We could consider returning the settings from the Join RPC (which replaces gossip-join), but let's hold off on that, at least in this PR, though it would be a good follow-up.

I assume this suggestion was motivated by not wanting to spend any time with stale settings? Instead of serializing settings through the Join RPC (which I think should stay simple - just node + cluster id + version, as it is today), perhaps what we could do here instead is wait for a gossip update containing the up-to-date settings. We'll want to do this after having joined the cluster, after having started up gossip, but before having fully started up the node. What do you think?

irfansharif reviewed (commented):
By the way, nice work here. Thanks for going through the review cycles; this is a much needed addition for crdb.
